pkg/scheduler/queue/queue_test.go

```diff
@@ -18,102 +19,126 @@ import (
-				err := queues[n].EnqueueRequest(users[j], requests[j], 0, nil)
-				if err != nil {
-					b.Fatal(err)
-				}
+func BenchmarkConcurrentQueueOperations(b *testing.B) {
+	req := "the query request"
+	maxQueriers := 0 // TODO: test with limit enforced?
+
+	for _, numTenants := range []int{1, 10, 1000} {
+		b.Run(fmt.Sprintf("%v tenants", numTenants), func(b *testing.B) {
+			for _, numProducers := range []int{10, 25} { // Query-frontends run 5 parallel streams per scheduler by default, and we typically see 2-5 frontends running at any one time.
+				b.Run(fmt.Sprintf("%v concurrent producers", numProducers), func(b *testing.B) {
+					for _, numConsumers := range []int{16, 160, 1600} { // Queriers run with parallelism of 8 when query sharding is enabled.
+						b.Run(fmt.Sprintf("%v concurrent consumers", numConsumers), func(b *testing.B) {
+							queueLength := promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"})
+							discardedRequests := promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"})
+							enqueueDuration := promauto.With(nil).NewHistogram(prometheus.HistogramOpts{})
+							queue := NewRequestQueue(100, 0, queueLength, discardedRequests, enqueueDuration)
+
+							start := make(chan struct{})
+							producersAndConsumers, ctx := errgroup.WithContext(context.Background())
+							require.NoError(b, queue.starting(ctx))
+							b.Cleanup(func() {
+								require.NoError(b, queue.stop(nil))
+							})
+
+							requestCount := func(total int, instanceCount int, instanceIdx int) int {
+								count := total / instanceCount
+								totalCountWithoutAdjustment := count * instanceCount
+
+								if totalCountWithoutAdjustment >= total {
+									return count
+								}
+
+								additionalRequests := total - totalCountWithoutAdjustment
+
+								if instanceIdx < additionalRequests {
+									// If we can't perfectly spread requests across all instances, assign remaining requests to the first instances.
+									return count + 1
+								}
+
+								return count
+							}
+
+							// TODO: limit the rate of production?
+							startProducer := func(producerIdx int) error {
+								requestCount := requestCount(b.N, numProducers, producerIdx)
+								tenantID := producerIdx % numTenants
+								<-start
+
+								for i := 0; i < requestCount; i++ {
+									for {
+										err := queue.EnqueueRequest(strconv.Itoa(tenantID), req, maxQueriers, func() {})
+										if err == nil {
+											break
+										}
+
+										// Keep retrying if we've hit the max queue length.
+										if !errors.Is(err, ErrTooManyRequests) {
+											return err
+										}
+									}
+
+									tenantID = (tenantID + 1) % numTenants
+								}
+
+								return nil
+							}
+
+							for producerIdx := 0; producerIdx < numProducers; producerIdx++ {
+								producerIdx := producerIdx
+								producersAndConsumers.Go(func() error {
+									return startProducer(producerIdx)
```

Comment:
a nitpick on naming: this runs the producer doesn't only start it

Reply:
same comment about the consumer function too

Reply:
Good point, changed in de00c70